internal/exec/stages/disks/disks.go

```diff
@@ -129,103 +125,117 @@ func (s stage) waitOnDevicesAndCreateAliases(devs []string, ctxt string) error {
 	return nil
 }
 
-// createPartitions creates the partitions described in config.Storage.Disks.
-func (s stage) createPartitions(config types.Config) error {
-	if len(config.Storage.Disks) == 0 {
+// createPartitionsAndRaids creates the partitions and raid devices described in config.Storage.Partitions
+// and config.Storage.Raid.
+func (s stage) createPartitionsAndRaids(config types.Config) error {
+	if len(config.Storage.Disks) == 0 && len(config.Storage.Raid) == 0 {
 		return nil
 	}
-	s.Logger.PushPrefix("createPartitions")
+	s.Logger.PushPrefix("createPartitionsAndRaids")
 	defer s.Logger.PopPrefix()
 
-	devs := []string{}
-	for _, disk := range config.Storage.Disks {
-		devs = append(devs, string(disk.Device))
-	}
-
-	if err := s.waitOnDevicesAndCreateAliases(devs, "disks"); err != nil {
-		return err
-	}
+	errChan := make(chan error)
+	// Mutex for preventing udev races by only doing one thing at a time. Also makes logs easier to read.
+	mutex := sync.Mutex{}
 
 	for _, dev := range config.Storage.Disks {
-		devAlias := util.DeviceAlias(string(dev.Device))
-
-		err := s.Logger.LogOp(func() error {
-			op := sgdisk.Begin(s.Logger, devAlias)
-			if dev.WipeTable {
-				s.Logger.Info("wiping partition table requested on %q", devAlias)
-				op.WipeTable(true)
-			}
-
-			for _, part := range dev.Partitions {
-				op.CreatePartition(sgdisk.Partition{
-					Number:   part.Number,
-					Length:   uint64(part.Size),
-					Offset:   uint64(part.Start),
-					Label:    string(part.Label),
-					TypeGUID: string(part.TypeGUID),
-					GUID:     string(part.GUID),
-				})
-			}
+		go s.createPartitions(errChan, &mutex, dev)
+	}
+	for _, dev := range config.Storage.Raid {
+		go s.createRaid(errChan, &mutex, dev)
+	}
 
-			if err := op.Commit(); err != nil {
-				return fmt.Errorf("commit failure: %v", err)
-			}
-			return nil
-		}, "partitioning %q", devAlias)
-		if err != nil {
+	// wait for all the goroutines to finish
+	for i := 0; i < len(config.Storage.Disks)+len(config.Storage.Raid); i++ {
+		if err := <-errChan; err != nil {
 			return err
```

Comment:
This can leak goroutines that are waiting for devices or the mutex. It could be handled but Ignition will be exiting shortly then anyway. Is it worth handling that or just let them be destroyed with the process when Ignition dies?

Reply:
There's no easy way to kill goroutines, so I say let them be leaked here and then destroyed when the program exits. Plus some of them could be endlessly stuck waiting on devices if one of them fails.